src/envoy/auth/jwt.cc

```diff
@@ -0,0 +1,168 @@
+#include "jwt.h"
+
+#include "common/common/base64.h"
+#include "common/common/utility.h"
+#include "openssl/evp.h"
+#include "openssl/rsa.h"
+#include "rapidjson/document.h"
+
+#include <algorithm>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace Envoy {
+namespace Http {
+namespace Auth {
+
+namespace Base64url {
+
+std::string decode(std::string input) {
+  std::replace(input.begin(), input.end(), '-', '+');
+  std::replace(input.begin(), input.end(), '_', '/');
+  switch (input.length() % 4) {
+    case 0:
+      break;
+    case 2:
+      input += "==";
+      break;
+    case 3:
+      input += "=";
+      break;
+    default:
+      /*
+       * for invalid base64url input, return empty string
+       */
+      return "";
+  }
+  return Base64::decode(input);
+}
+
+}  // namespace Base64url
+
+namespace Util {
+
+const uint8_t* unsignedCStr(const std::string& str) {
+  return reinterpret_cast<const uint8_t*>(str.c_str());
+}
+
+}  // namespace Util
+
+bssl::UniquePtr<EVP_PKEY> Jwt::evpPkeyFromStr(const std::string& pkey_pem) {
+  std::string pkey_der = Base64::decode(pkey_pem);
+  bssl::UniquePtr<RSA> rsa(RSA_public_key_from_bytes(
+      Util::unsignedCStr(pkey_der), pkey_der.length()));
+  if (rsa == nullptr) {
```

Comment:
[`bssl::UniquePtr<T>`](https://github.com/google/boringssl/blob/master/include/openssl/base.h#L439)
is `std::unique_ptr` with specified deleter (e.g. `RSA_free()` in this case)